File: javaFile/reverse_Sub_Part.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    LinkedList() {
        this.head = null;
    }

    void insert(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = newNode;
        }
    }

    void reverseSubPart(Node head, int start, int end) {
        Node current = this.head, prev_m = null,
        mth = null, nth = null, after_n = null;
        for (int i = 1; current.next != null; i++, current = current.next) {
            if (i == start - 1) {
                prev_m = current;
            }
            if (i == start) {
                mth = current;
            }
            if (i == end) {
                nth = current;
                after_n = current.next;
                break;
            }
        }

    }
}

File: python/reverseSubPart.py
class Node:
    def __init__(self, data):
        self.data = None
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def push(self, new_data):
        new_node = Node(new_data)
        new_node.next = self.head
        self.head = new_node

    def reverse_sub_part(self, start, end):
        if start == end:
            return

        prev = None
        current = self.head

        # Move to the start position
        for _ in range(start - 1):
            prev = current
            current = current.next

        last_node_of_first_part = prev
        last_node_of_sub_list = current

        # Reverse the sub-list
        next_node = None
        for _ in range(end - start + 1):
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        if last_node_of_first_part:
            last_node_of_first_part.next = prev
        else:
            self.head = prev

        last_node_of_sub_list.next = current

    def print_list(self):
        current = self.head
        while current:
            print(current.data, end=" ")
            current = current.next
        print()

def main():
    llist = LinkedList()
    llist.push(5)
    llist.push(4)
    llist.push(3)
    llist.push(2)
    llist.push(1)

    print("Original Linked List:")
    llist.print_list()

    start = 2
    end = 4
    llist.reverse_sub_part(start, end)

    print(f"Reversed sub-part from {start} to {end}:")
    llist.print_list()

if __name__ == "__main__":
    main()
    